src/main/java/software/amazon/smithy/lsp/ext/SmithyProject.java

```diff
@@ -159,37 +161,162 @@ public File getRoot() {
-            String fileName = sourceLocation.getFilename();
-            String uri = Utils.isJarFile(fileName)
-                ? Utils.toSmithyJarFile(fileName)
-                : !fileName.startsWith("file:") ? "file:" + fileName
-                : fileName;
+    private static Map<ShapeId, Location> collectLocations(Model model) {
+        Map<ShapeId, Location> locations = new HashMap<>();
+        List<String> modelFiles = model.shapes()
+                .map(shape -> shape.getSourceLocation().getFilename())
+                .distinct()
+                .collect(Collectors.toList());
+        for (String modelFile : modelFiles) {
+            List<String> lines = getFileLines(modelFile);
+            int endMarker = getInitialEndMarker(lines);
+            int memberEndMarker = getInitialEndMarker(lines);
+
+             // Get shapes reverse-sorted by source location to work from bottom of file to top.
+            List<Shape> shapes = model.shapes()
+                    .filter(shape -> shape.getSourceLocation().getFilename().equals(modelFile))
+                    // TODO: Once the change in https://github.com/awslabs/smithy/pull/1192 lands, replace with with
+                    // `.sorted(Comparator.comparing(Shape::getSourceLocation).reversed())`.
+                    .sorted(new SourceLocationSorter().reversed())
+                    .collect(Collectors.toList());
 
-            Position pos = new Position(sourceLocation.getLine() - 1, sourceLocation.getColumn() - 1);
-            Location location = new Location(uri, new Range(pos, pos));
 
-            String shapeName = shape.getId().getName();
-            // Members get the same shapeName as their parent structure
-            // so we ignore them, to avoil producing a location per-member
-            // TODO: index members somehow as well?
-            if (shape.getType() != ShapeType.MEMBER) {
-                if (locations.containsKey(shapeName)) {
-                    locations.get(shapeName).add(location);
+            for (Shape shape : shapes) {
+                SourceLocation sourceLocation = shape.getSourceLocation();
+                Position startPosition = new Position(sourceLocation.getLine() - 1, sourceLocation.getColumn() - 1);
+                Position endPosition;
+                if (endMarker < sourceLocation.getLine()) {
+                    endPosition = new Position(sourceLocation.getLine() - 1, sourceLocation.getColumn() - 1);
                 } else {
-                    List<Location> locList = new ArrayList<Location>();
-                    locList.add(location);
-                    locations.put(shapeName, locList);
+                    endPosition = getEndPosition(endMarker, lines);
                 }
-            }
-        });
 
+                // Find the end of a member's location by first trimming trailing commas, empty lines and closing
+                // structure braces.
+                if (shape.getType() == ShapeType.MEMBER) {
```

Comment:
Is it reasonable to break these branches up in to their own methods for readability?